service/src/main/java/bio/terra/cbas/runsets/results/RunCompletionHandler.java

```diff
@@ -68,93 +68,123 @@ public void updateOutputAttributes(Run run, Object outputs)
       }
 
       // Pull workflow completion information and save run outputs
       ArrayList<String> errors = new ArrayList<>();
 
-      // Saving run outputs for a Successful status only.
-      if (updatedRunState == CbasRunStatus.COMPLETE && workflowOutputs != null) {
+      // Saving run outputs for a Successful terminal status only.
+      if (updatedRunState == CbasRunStatus.COMPLETE) {
+
         // Assuming that if no outputs were not passed with results,
         // Then no explicit pull should be made for outputs from Cromwell.
-        String errorMessage = saveWorkflowOutputs(updatableRun, workflowOutputs);
-        if (errorMessage != null && !errorMessage.isEmpty()) {
-          // If this is the last attempt to save workflow outputs,
-          // update run info in database with an error.
-          // Otherwise, we should return internal error to caller.
-          // TODO - handling OutputProcessingException as User Error will be covered in [WM-2090]
+        try {
+          boolean processedSuccessfully = saveWorkflowOutputsToWds(updatableRun, workflowOutputs);
+          if (!processedSuccessfully) {
+            return RunCompletionResult.VALIDATION;
+          }
+        } catch (Exception e) {
+          // log WDS or other Runtime error and mark Run as Failed.
+          String errorMessage =
+              "Error while updating data table attributes for record %s from run %s (engine workflow ID %s): %s"
+                  .formatted(
+                      updatableRun.recordId(),
+                      updatableRun.runId(),
+                      updatableRun.engineId(),
+                      e.getMessage());
+          logger.error(errorMessage, e);
           errors.add(errorMessage);
+        }
+
+        if (!errors.isEmpty()) {
+          // Update run info in database with an error.
           updatedRunState = CbasRunStatus.SYSTEM_ERROR;
         }
       } else if (updatedRunState.inErrorState()) {
         if (workflowErrors != null && !workflowErrors.isEmpty()) {
           errors.addAll(workflowErrors);
         }
       }
-
       // Save the updated run record in database.
-      updateResult = updateDatabaseRunStatus(updatableRun, updatedRunState, errors);
+      updateResult =
+          updateDatabaseRunStatus(updatableRun, updatedRunState, errors, engineStatusChange);
     } finally {
       recordMethodCompletion(updateResultsStartNanos, RunCompletionResult.SUCCESS == updateResult);
     }
     // we should not come here
     return updateResult;
   }
 
-  /*
-   The method checks if output definitions are associated with run and makes updates.
-   If outputs passed as arguments are null, but the workflow has the output definitions,
-   then call to Cromwell will be made to pull the outputs.
-  */
-  private String saveWorkflowOutputs(Run updatableRun, Object outputs) {
+  private boolean saveWorkflowOutputsToWds(Run updatableRun, Object workflowOutputs)
```

Comment:
And can we maybe add a good set of unit tests for this function? It feels like a great place to make sure we're doing the right thing for all the wide variety of possible input arguments for `Run` and `Object`?

Reply:
And thinking out loud about making unit testing easier, I wonder whether it would make sense to break this function into two halves - one which is more like a functional mapping that takes these inputs and works out, as a batch, what WDS updates to make (which should be easy to unit test), and then a second half which turns all those WDS updates into actual API calls (which can probably just be a single simple test)? Just a thought!